plaidml/bridge/openvino/ops/deformable_convolution.cpp

```diff
@@ -0,0 +1,248 @@
+  deformed_index_reshape_dims.insert(deformed_index_reshape_dims.end(), deformed_dims.begin(), deformed_dims.end());
+  deformed_index_reshape_dims.push_back(1);
+  auto deformed_index_slice = op::slice(deformed_index).add_dim(0, N).add_dim(0, CI);
+  for (auto i = 0; i < rank - 2; ++i) {
+    deformed_index_slice = op::slice(deformed_index_slice).add_dim(0, deformed_dims[i]);
+  }
+  for (auto i = 0; i < rank - 2; ++i) {
+    deformed_index_vec[i] = op::slice(deformed_index_slice).add_dim(i);
+    deformed_index_vec_ceil[i] = edsl::ceil(deformed_index_vec[i]);
+    deformed_index_vec_floor[i] = edsl::floor(deformed_index_vec[i]);
+    deformed_index_vec_ceil[i] = op::reshape(deformed_index_vec_ceil[i], make_tuple<int>(deformed_index_reshape_dims));
+    deformed_index_vec_floor[i] =
+        op::reshape(deformed_index_vec_floor[i], make_tuple<int>(deformed_index_reshape_dims));
+  }
+  // Coord_num is the number of tensor required for multi-linear interpolation.
+  int coord_num = static_cast<int>(std::pow(2.0, rank - 2));
+  std::vector<edsl::Tensor> deformed_index_coord_vec(coord_num);
+  std::vector<edsl::Tensor> deformed_input_vec(coord_num);
+  for (auto i = 0; i < coord_num; ++i) {
+    std::vector<edsl::Tensor> concat_vec(rank - 2);
+    for (auto j = 0; j < rank - 2; ++j) {
+      int flag = (i / static_cast<int>(std::pow(2.0, rank - 2 - j - 1))) % 2;
+      concat_vec[j] = (flag == 0) ? deformed_index_vec_floor[j] : deformed_index_vec_ceil[j];
+    }
+    deformed_index_coord_vec[i] = op::concatenate(concat_vec, rank);
+    deformed_index_coord_vec[i] = edsl::cast(deformed_index_coord_vec[i], DType::INT32);
+    deformed_input_vec[i] = edsl::gather(I, deformed_index_coord_vec[i]).mode(edsl::GatherMode::ND).batchDims(2);
+  }
+  // Change the shape of deformed_index_ceil tensor and deformed_index_floor tensor to original shape.
+  // For example, in 2D, after this operation, the shapes of deformed_index_ceil tensor and deformed_index_floor tensor
+  // will be {N, CI, OFF_H*F_H, OFF_W*F_W}.
+  deformed_index_reshape_dims.erase(deformed_index_reshape_dims.end() - 1);
+  for (auto i = 0; i < rank - 2; ++i) {
+    deformed_index_vec_ceil[i] = op::reshape(deformed_index_vec_ceil[i], make_tuple<int>(deformed_index_reshape_dims));
+    deformed_index_vec_floor[i] =
+        op::reshape(deformed_index_vec_floor[i], make_tuple<int>(deformed_index_reshape_dims));
+  }
+  // Get deformed input tensor with multi-linear interpolation.
+  for (auto i = 0; i < rank - 2; ++i) {
+    for (auto j = 0; j < static_cast<int>(std::pow(2.0, rank - 2 - i - 1)); ++j) {
+      deformed_input_vec[j] = deformed_input_vec[j] * (deformed_index_vec_ceil[i] - deformed_index_vec[i]) +
+                              deformed_input_vec[j + static_cast<int>(std::pow(2.0, rank - 2 - i - 1))] *
+                                  (1.0 - deformed_index_vec_ceil[i] + deformed_index_vec[i]);
+    }
+  }
+  edsl::Tensor deformed_input = deformed_input_vec[0];
```

Comment:
Is only deformed_input_vec[0] used? so why we calculate every element of deformed_input_vec?

Reply:
deformed_input_vec[0] is calculated by every element of deformed_input_vec, and final deformed_input_vec[0] is different from the origin deformed_input_vec[0].

Reply:
for example, in 3D convolution, the size of deformed_input_vec is 8, and the origin deformed_input_vec is assigned by the result gather ND, and the order is sorted according Binary number， then first calculation, the former 4 element is calculated by the whole deformed_input_vec, the second calculation, the former 2 element is calculated by the former 4 element, the third calculation, the former 1 element is calculated by former 2 element.

Reply:
Thanks, I see.

Reply:
Group size is not used now?

Reply:
I just throw exception when group is not equal 1 according to the tim's review.